vignettes/loo2-mixis.Rmd

```diff
@@ -0,0 +1,198 @@
+
+```{r, results='hide', warning=FALSE}
+chains <- 4
+n_iter <- 2000
+warm_iter <- 1000
+stanmodel <- stan_model(model_code = stancode_horseshoe)
+fit_post <- sampling(stanmodel, data = standata, chains = chains, iter = n_iter, warmup = warm_iter, refresh = 0)
+loo_post <-loo(fit_post)
+```
+
+```{r}
+print(loo_post)
+```
+
+As we can see the diagnostics signal either "bad" or "very bad" Pareto-$k$ values for roughly $15-30\%$ of the observations which is a significant portion of the dataset. 
+
+## Mixture estimators
+
+We now compute the mixture estimators proposed in Silva and Zanella (2022). These require to sample from the following mixture of leave-one-out posteriors
+\begin{equation}
+q_{mix}(\theta) =  \frac{\sum_{i=1}^n p(y_{-i}|\theta)p(\theta)}{\sum_{i=1}^np(y_{-i})}\propto p(\theta|y)\cdot \left(\sum_{i=1}^np(y_i|\theta)^{-1}\right).
+\end{equation}
+The code to generate a Stan model for the above mixture distribution is the same to the one for the posterior, just enabling one line of code with a _LogSumExp_ contribution to account for the last term in the equation above.
+```
+  if (mixis) {
+    target += log_sum_exp(-log_lik);
+  }
+```
+
+We sample from the mixture and collect the log-likelihoods term.
+```{r, results='hide', warnings=FALSE}
+standata$mixis <- 1
+fit_mix <- sampling(stanmodel, data = standata, chains = chains, iter = n_iter, warmup = warm_iter, refresh = 0, pars = "log_lik")
+log_lik_mix <- extract(fit_mix)$log_lik
+```
+
+We now compute the mixture estimators, following the numerically stable implementation in Appendix A.2 of [Silva and Zanella (2022)](https://arxiv.org/abs/2209.09190). The code below makes use of the package "matrixStats".
+
+```{r}
+l_common_mix <- rowLogSumExps(-log_lik_mix)
+log_weights <- -log_lik_mix - l_common_mix
+elpd_mixis <- logSumExp(-l_common_mix) - rowLogSumExps(t(log_weights))
+```
+
+## Comparison with benchmark values obtained with long simulations
+
+To evaluate the performance of mixture estimators (MixIS) we also generate _benchmark values_, i.e.\ accurate approximations of the LOO predictives $\{p(y_i|y_{-i})\}_{i=1,\dots,n}$, obtained by brute-force sampling from the leave-one-out posteriors directly, getting $90k$ samples from each and discarding the first $10k$ as warmup. This is computationally heavy, hence we have saved the results and we just load them in the current vignette. 
+```{r}
+data(voice_loo)
+elpd_loo <- voice_loo$elpd_loo
+```
+
+We can then compute the root mean squared error (RMSE) of the PSIS and mixture estimators relative to such benchmark values.
+```{r}
+elpd_psis <- loo_post$pointwise[,1]
+print(paste("RMSE(PSIS) =",round( sqrt(mean((elpd_loo-elpd_psis)^2)) ,2)))
+print(paste("RMSE(MixIS) =",round( sqrt(mean((elpd_loo-elpd_mixis)^2)) ,2)))
```

Comment:
When I run this these come out as NaN. It looks like `voice_loo` has columns "X" and "bk_loo" but no column `elpd_loo`. So `voice_loo$elpd_loo` comes out to be NULL.

Reply:
should be fixed now

Reply:
Thanks! I'll take a look soon